### ISSUE TITLE
Integration tests

### ISSUE THREAD

**Issue by adomasven** [zotero/zotero-connectors#135]
This is becoming pretty critical now. We can get away without unit tests, but we must ensure basic functionality doesn't accidentally break.

Platforms:
- [ ] Chrome (Latest)
- [ ] Firefox (Latest)
- [ ] Firefox (ESR)
- [ ] Safari (Latest)
- [ ] Safari (Oldest-supported)

Having at least Firefox ESR would be a good start.

Functionality:
- [ ] Debug output submission
- [ ] Report ID submission
- [ ] Translation to Zotero
- [ ] Translation to Server
- [ ] Prompts
- [ ] Notifications
- [ ] Proxy detection
- [ ] Style installation
- [ ] Data-files import

**Comment by adomasven** [comment]
So before I begin, this is the plan:
- Use the [node.js Selenium Webdriver](http://seleniumhq.github.io/selenium/docs/api/javascript/index.html) to drive browsers
- Use the [Mocha](https://mochajs.org/)-[Chai](http://chaijs.com/) testing framework (currently used by the Zotero client unit-test suite)
- Use [Express.js](https://expressjs.com/) to mock responses from the client, zotero.org and translation targets.

I'll start with Firefox ESR. Any objections/suggestions @dstillman @tnajdek @inukshuk

**Comment by adomasven** [comment]
The plan has changed to containing the tests within the extension with the debugging build. The webdriver spec and implementations may be ready in 5 years to build something reasonable, but not today. We might still use Selenium to actually launch the browsers with extensions preloaded, but that's not strictly necessary. Express.js had been ditched too as we can stub responses with sinon.js since the tests run in the inside the blackbox.

**Comment by adomasven** [comment]
Safari automation is stuck [upstream](https://github.com/travis-ci/travis-ci/issues/7187). High Sierra (release in 2 weeks) required to launch a session with automation enabled and then it will probably take a while for travis-ci to support it.

Either way, I'll work on tests and we can at least run them manually before releases, which will still catch bugs.

**Comment by adomasven** [comment]
Seems like we might be able to set up travis automation now, see upstream link above.

**Comment by adomasven** [comment]
We should consider rewriting these since they rely on eval that is no longer viable with MV3, but there have been improvements to headless browsers and selenium since writing this that would allow to automate the browser and get its contents directly in nodejs.

**Comment by mbrukman** [comment]
Hi @adomasven, it appears this repo is still using Travis CI, given the presence of the [`.travis.yml`](https://github.com/zotero/zotero-connectors/blob/04cdbb8b988710f7bff4a2daf79bcd73c45fffe2/.travis.yml) config file, but this project appears to no longer be registered with Travis CI, as visiting https://app.travis-ci.com/github/zotero/zotero-connectors says:

> This is not an active repository

Would it make sense to migrate to GitHub Actions? I see the main repo https://github.com/zotero/zotero is using GitHub Actions rather than Travis CI.

**Comment by adomasven** [comment]
Yes, overall the timing of our adding these tests was not great. Browser automation just wasn't ready in 2017 and then the announcement of Manifest V3 meant that we couldn't continue using the implementation we had, although Manifest V3 itself was then delayed all the way until 2023. Anyway, the tests need to be rewritten from scratch, and we will use github actions when we do.
